### DISCUSSION TITLE
Custom Fonts

### DISCUSSION THREAD

**Discussion by teeshq** [lovyan03/LovyanGFX#321] [Q&A]
Hi how to use custom fonts where i can paste them ect

**Comment by martinberlin** [comment]
I have the feeling that this part is missing some documentation @lovyan03 

I got stuck for example in the latest days trying to render text in Polish, which seems that Adafruit GFX **fontconvert** tool does not support. As an example:

> GADZIE GODY   czyli Sagi o czaroksiążcy prapoczątek, opowieść o tym jak młody Debren na praktykę magiczną do wrogiej krainy posłan oko w oko ze smokiem staje i utraconą cześć dziewiczą przywrócić musi     - Zarooobki dla dziewic! - przebił się przez uliczny gwar basowy okrzyk. - Gooodziwe zarooobki! Debren potknął się z wrażenia, a potem przyspieszył kroku.

**Comment by tobozo** [comment]
@martinberlin I know how you feel

![image](https://user-images.githubusercontent.com/1893754/205047120-9b135374-3a3b-4f28-a679-074c997260d6.png)

The [font example sketch](https://github.com/lovyan03/LovyanGFX/blob/master/examples/HowToUse/3_fonts/3_fonts.ino) should help you with the declarative part.

Here's another [example](https://github.com/riraosan/ESP32_8BIT_CVBS/blob/master/sample/04_unicode_fonts/u8g2/u8g2.ino) on how to import u8g2 unicode fonts into lgfx.

Contributions to the wiki of this repository are welcome btw.

**Comment by martinberlin** [comment]
Thanks!
Lovyan also told me is possible to generate Fonts using https://github.com/m5stack/FontCreator so I will try with that as well just to compare the results.
[Video of my 9.7" epub reader DIY build using LovyanGFX](https://twitter.com/martinfasani/status/1597256098301374465)

**Comment by martinberlin** [comment]
https://github.com/m5stack/FontCreator -> Discarded for me. Bad documented. Generating all Ubuntu fonts also does not render polish. I don't really see much of a difference with Adafruit GFX font converter.
So far lass test will be with **u8g2** and then if nothing works will do it with Open Font Render, that it worked already for me when doing a Chinese weather panel.

**Comment by tobozo** [comment]
there's still a way to load font files from the filesystem with `tft.loadFont( "/path/to/ttf/file", SPIFFS )`

found [another documentation](https://github.com/lovyan03/LovyanGFX/tree/master/src/lgfx/Fonts/IPA) based on u8g2

- [ttf to bdf](http://sofia.nmsu.edu/~mleisher/Software/otf2bdf/)
- [bdf to c](https://github.com/olikraus/u8g2/blob/master/tools/font/bdfconv/bdfconv.exe)

bdfconv settings for polish (found in [u8g2 MakeFile](https://github.com/olikraus/u8g2/blob/7e58b87b3cd63ed4016624cd00226ffe4f0bfe1d/tools/font/bdfconv/Makefile#L66)): 

	./bdfconv -f 1 -v -m '32-127,$$104-$$107,$$118-$$119,$$141-$$144,$$15a-$$15b,$$179-$$17c,$$d3,$$f3' ../bdf/unifont.bdf -n unifont_polish -o unifont_polish.c -d ../bdf/7x13.bdf

**Comment by martinberlin** [comment]
Nice @tobozo thanks a lot. That of storing the ttf in SPIFFS sounds quite useful must say.
Would like to know, the way to use it will be to do something like this?

```C
GFXfont *m_regular_font = tft.loadFont( "/path/to/ttf/file", SPIFFS);
```

Does it return a GFXfont type or it's something different. 
If I could get a complete working example I will try it today. Thanks a lot for the hints!

**Comment by tobozo** [comment]
sorry it's not TTF but VLW fonts, you can generate those using the `Processing` software.

There's a [tutorial on seeed wiki](https://wiki.seeedstudio.com/Wio-Terminal-LCD-Anti-aliased-Fonts/).

Processing will only see the system fonts though (no file picker), so make sure to install them before launching the interface.

And it's just this, tft.loadFont() has no return value:

`tft.loadFont( "/my_VLW_font.vlw", SPIFFS);`

I just tried it with `AbhayaLibre-Medium-48` on LittleFS and got this result, the font is downscaled at 50% so don't mind the quality loss:

![image](https://user-images.githubusercontent.com/1893754/205279650-0515b733-3d11-4891-aa7c-7e4bd38f3402.png)

**Comment by martinberlin** [comment]
Great to see that it works. I will try that then!
Can you please copy the C part where you load the font and then assign it to the display?

I still don't see how to do a:   display.setFont(&MyFont);

**Reply by tobozo** [reply]
```cpp


void setup()
{
  Serial.begin(115200);
  LittleFS.begin();
  Serial.println("ESP32 Ready");
  
  tft.init();

  tft.loadFont("/AbhayaLibre-Medium-48.vlw", LittleFS);

  tft.setTextSize(0.5);

  tft.println();

  tft.println("    ąćęłńóśźż    ");
  tft.println("    ĄĆĘŁŃÓŚŹŻ    ");

  tft.setTextSize(1);
}
```

**Reply by martinberlin** [reply]
That looks like a great solution. The problem for the ePub project is that when rendering HTML you need sometimes to switch from one Regular to Bold or similar, and I think doing it this way it will have kind of a great overhead.
But if you use it as default font for everything I guess this is the way to go!
I will give it a try also to U82G. And then also maybe tweaking fontconvert (Adafruit GFX which stops at 0x7E 126
![Screenshot from 2022-12-02 11-01-28](https://user-images.githubusercontent.com/2692928/205300900-2d80bad8-acc4-4607-a702-540dd20d1e3c.png)
) to support 8-bit font Glyphs (Why they limited this to 7-bit?) or to also tweak https://github.com/vroland/epdiy/blob/master/scripts/fontconvert.py to also generate GLXFont format since it already supports 8-bit definitions.

**Reply by martinberlin** [reply]
Regarding this @tobozo 

       tft.loadFont("/AbhayaLibre-Medium-48.vlw", LittleFS);

I don't think this is an option when using IDF. Because I checked the code and over those functions there is always an:

     #ifdef ARDUINO
     
Which won't compile that part for me. I think there is another one available that is:

     void loadFont(const char *path)

But it won't work as is with a path, unless I have a way to tell it that it's SPIFFS. Or I should find a way to preload each of this fonts in SPIFFs, which does not seem an option. At the end I've loosed a lot of time with this that could be used to implement Open Font Render, which works and it seems like a more straight-forward option. But thanks a lot to make me discover this new functions!

**Comment by martinberlin** [comment]
Got it working using [latest **fontconvert** from Adafruit](https://github.com/martinberlin/cale-idf/commit/8330f9b35c8be63d77887468a06228739fb238cf)

Example on how to do it:

     /fontconvert /usr/share/fonts/truetype/ubuntu/Ubuntu-R.ttf 12 32 350 > Ubuntu_R12pt8b.h

**Note:** It seems the polish glyphs are in the area that is from 255 -> 350 (Later will open the font with FontForge and check more specifically)

**Reply by lefty01** [reply]
Hi, may I jump in on this, I need some font with german umlaut. I figured I might be able to use /usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf for example, convert with:
$ ./fontconvert /usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf 12 32 350 > LiberationMono-Regular.h

but still note sure how can I use / load the font.
could u please post the snippet where you use the include file and call setFont? thx.

**Reply by lefty01** [reply]
ah ... ok it's just like that, include the header (output from fontconvert) and use setFont.
run (check here for example for number of last char to include https://symbl.cc/en/unicode/table/)
`$ ./fontconvert LiberationMono-Regular.ttf 18  32 382  > LiberationMono-Regular_18.h`
in code:
```
#include "LiberationMono-Regular_18.h"
...
  lcd.setFont(&LiberationMono_Regular18pt8b);
  lcd.println("Hellö Wüorlß");
  lcd.println("123 €");

```

**Comment by martinberlin** [comment]
Thanks a lot @tobozo 
[Polish font render preview](https://twitter.com/martinfasani/status/1598768769492393984) plus idea I got from seeing Bodmer update in Open Font Render
